### ISSUE TITLE
DOSBox-X crashes when attempting to start Command & Conquer installer

### ISSUE THREAD

**Issue by hazelnot** [joncampbell123/dosbox-x#3285]
### Code of Conduct & Contributing Guidelines

- [x] I agree to follow the code of conduct and the contributing guidelines.

### Have you checked that no other similar bug report(s) already exists?

- [X] I have searched and didn't find any similar issues.

### What operating system(s) this bug have occurred on?

Arch Linux

### What version(s) of DOSBox-X have this bug?

0.83.22 Linux SDL1 64-bit

### Describe the bug

When attempting to start the Command & Conquer installer from either the GDI or Nod discs, DOSBox-X simply crashes.

### Steps to reproduce the behaviour

1. Mount the ISO, either directly from DOSBox or from the host OS
2. Switch to the drive letter where the folder/ISO is mounted
3. Start `setup.exe`
4. Crash

### Used configuration

[dosbox-x.conf.txt](https://github.com/joncampbell123/dosbox-x/files/8097021/dosbox-x.conf.txt)

### Emulator log

```text
LOG: Early LOG Init complete
LOG: DOSBox-X's working directory: /home/hazelnot
LOG: Logging init: beginning logging proper. This is the end of the early init logging
LOG: Logging: No logfile was given. All further logging will be discarded.
LOG: DOSBox-X version 0.83.22 (Linux SDL1)
LOG: Current X11 keyboard layout (token) is: 'us'
LOG: Host keyboard layout is now us (US English)
LOG: Mapper keyboard layout is now us (US English)
LOG: SDL: TTF activated 
LOG: Configured windowposition: 
LOG: X11 main window is 640 x 400 maximized=0
LOG: X11 extension XRANDR is available
LOG: XRandR CRTC 0: pos=(0,0) size=(2560,1440) outputs=1
LOG: Our window lies on this CRTC display (window pos=(140,174) size=(640,400) match=(460,374)).
LOG:   Goes to output 0: name='XWAYLAND0' size_mm=(600 x 340)
LOG: Screen report: Method 'XRandR' (2560.000 x 1440.000 pixels) at (0.000 x 0.000) (600.000 x 340.000 mm) (23.622 x 13.386 in) (108.373 x 107.576 DPI)
LOG: ISA BCLK: 8333333.333Hz (25000000/3)
LOG: monopal: green, 
LOG: Active save slot: 1 [Empty]
LOG: USING AVI+ZMBV
LOG: Max 3670016 sz 16384
LOG: Final 16384
LOG: Glide:LFB access: read-write (no aux)
LOG: Glide:Unable to load glide2x library, glide emulation disabled
LOG: MIDI:Opened device:oss
LOG: Pentium CMPXCHG8B emulation is enabled
LOG: VOODOO LFB now at d0000000
LOG: Serial1: BASE 3f8h
LOG: Serial2: BASE 2f8h
LOG: Parallel1: BASE 378h
LOG: MPU-401 Registering I/O ports as if IBM PC MPU-401 at base 330h
LOG: X11 main window is 640 x 400 maximized=0
LOG: XRandR CRTC 0: pos=(0,0) size=(2560,1440) outputs=1
LOG: Our window lies on this CRTC display (window pos=(140,174) size=(640,400) match=(460,374)).
LOG:   Goes to output 0: name='XWAYLAND0' size_mm=(600 x 340)
LOG: Screen report: Method 'XRandR' (2560.000 x 1440.000 pixels) at (0.000 x 0.000) (600.000 x 340.000 mm) (23.622 x 13.386 in) (108.373 x 107.576 DPI)
LOG: Allocated APM BIOS pm entry point at f000:ce20
LOG: Writing code to fce20
LOG: X11 main window is 640 x 400 maximized=0
LOG: XRandR CRTC 0: pos=(0,0) size=(2560,1440) outputs=1
LOG: Our window lies on this CRTC display (window pos=(140,174) size=(640,400) match=(460,374)).
LOG:   Goes to output 0: name='XWAYLAND0' size_mm=(600 x 340)
LOG: Screen report: Method 'XRandR' (2560.000 x 1440.000 pixels) at (0.000 x 0.000) (600.000 x 340.000 mm) (23.622 x 13.386 in) (108.373 x 107.576 DPI)
LOG: ISA Plug & Play BIOS enabled
LOG: VGA ROM BIOS init callback
LOG: pixratio 1.350, dw false, dh false
LOG: Aspect ratio: 720 x 540  xToY=1.333 yToX=0.750
LOG: SDL: TTF activated 
LOG: SDL: TTF activated 
LOG: SDL: TTF activated 
LOG: WARNING: No translation support (to host) for code page 0
LOG: Loaded system codepage: 437
LOG:    1861577 ERROR BIOS:Keyboard layout file auto not found
LOG:    1861577 ERROR BIOS:Keyboard layout file auto not found
LOG: XMS: 50 handles allocated for use by the DOS environment
LOG: EMS page frame at 0xe000-0xefff
LOG: COMMAND.COM env size:             720 bytes
LOG: COMMAND.COM environment block:    0x0701 sz=0x002d
LOG: COMMAND.COM main body (PSP):      0x072f sz=0x009a
LOG: COMMAND.COM stack:                0x0749
*** buffer overflow detected ***: terminated
zsh: IOT instruction (core dumped)  dosbox-x
```

### Additional info

Starting the setup program from Windows 3.1 actually worked, oddly enough, although when trying to start the game from either Windows or DOS I get a black screen

**Comment by hazelnot** [comment]
Possibly related, I then tried to install Red Alert, failed, installed it from regular DOSBox, and tried to start it from DOSBox-X, and the game started but failed to detect a CD drive.

**Comment by DSN9990** [comment]
I just tried it right now and it seems to work fine too, game appears to install and boot fine here as well. No problems so far. I have also added the "Covert Operations" missions to my setup too.

**Comment by grapeli** [comment]
I am using archlinux, I am using dosbox-x, and I have no problems installing and running c&c here.

https://user-images.githubusercontent.com/452325/154975582-e6e75fee-2574-4500-b9e1-88988e03b9c0.mp4

**Comment by rderooy** [comment]
I also tested it on Fedora Linux 35 and the installer worked fine.

**Comment by hazelnot** [comment]
That's... weird. Is there anything else I could provide to figure out why it's specifically not working for me?

**Comment by grapeli** [comment]
Enter the given phrase in the search engine - "IOT instruction (core dumped)".

Maybe you compiled dosbox-x on different hardware and run it on a different (older) cpu.
Use gdb. Because the information you provided is too meager to draw conclusions.

**Comment by maron2000** [comment]
Setup crashed (buffer overflowed) on Ubuntu 18.04.
Try `loadfix -a` then `setup`.
Seems to work for me.

**Comment by grapeli** [comment]
I didn't add the `loadfix -a` command and setup, installation and gameplay work fine. Both with version 1.07 and after updating to 1.22.
The dosbox-x configuration is practically the default.

https://user-images.githubusercontent.com/452325/155103832-b3022297-576b-4cff-878e-4815f987f108.mp4

**Comment by rderooy** [comment]
I tried it with the SDL1 and SDL2 DOSBox-X version with v1.22 of the game. No need for loadfix and a default config file is able to launch the installer just fine.

**Comment by grapeli** [comment]
I checked with the linux dosbox-x version based on sdl1.x (I don't use it completely). C&C works perfectly fine in it.

**Comment by hazelnot** [comment]
> Enter the given phrase in the search engine - "IOT instruction (core dumped)".
> 
> Maybe you compiled dosbox-x on different hardware and run it on a different (older) cpu. Use gdb. Because the information you provided is too meager to draw conclusions.

I compiled it on this computer, I use the AUR version which compiles locally.

**Comment by grapeli** [comment]
The gdb output is represented by these lines. Not very important.

```
[Detaching after vfork from child process 10860]
[Detaching after vfork from child process 11108]
[Detaching after vfork from child process 11136]

Thread 1 "dosbox-x" received signal SIGABRT, Aborted.
0x00007f66712ad34c in __pthread_kill_implementation () from /usr/lib/libc.so.6
```

After all, @maron2000  confirmed crash under ubuntu for the sdl1 version.
He handed out the finished solution. **You must add the** `loadfix -x` **command before running** `setup.exe`.
Follow this suggestion. Inform if it works effectively.

**Comment by hazelnot** [comment]
Oh. Someone said the loadfix was unnecessary so I didn't try it earlier. I tried it now and now it works. I think I'm gonna switch to the SDL2 version though, since I wasn't aware that existed until now.

**Comment by grapeli** [comment]
@hazelnot 
I do not use AUR. First, it is not part of archlinux. Lots of crap there, prepared completely by incompetent persons. You use it entirely at your own risk. It carries a high risk, especially in the case of inexperienced users. Unknowingly, they make a leaky Swiss cheese out of the operating system.

I checked this [PKBUILD](https://aur.archlinux.org/packages/dosbox-x) dosbox-x that you probably used. In fact, now I can confirm this crash as well. The cause of the crash is **hardening compiler flags** (`-Wp,-D_FORTIFY_SOURCE=2`).
If you add the following lines to the PKBUILD in the build() section then this crash does not occur.
```
  export CFLAGS=${CFLAGS/-Wp,-D_FORTIFY_SOURCE=2/}
  export CXXFLAGS=${CXXFLAGS/-Wp,-D_FORTIFY_SOURCE=2/}

```(before the ending line with `./build`)

**Comment by maron2000** [comment]
@grapeli So the problem is probablly due to glitched access to the emulated main memory, I tried changing the options below and it worked on Ubuntu.
I assume the treatment of `mcb free` is the key for this bug.
Anyway, `loadfix` is acceptable as long as there are sufficient free memory for the game.
```
minimum mcb segment = 500
minimum mcb free = 1
```

**Comment by hazelnot** [comment]
> I do not use AUR (completely). First, it is not part of archlinux. Lots of crap there, prepared completely by incompetent persons. You use it entirely at your own risk. It carries a high risk, especially in the case of inexperienced users. Unknowingly, they make a leaky Swiss cheese out of the operating system.

Thank you for your help, but that's what I came here for, help, not elitism and condescension. The AUR _is_ part of Arch, whether you like it or not, and I do know what I'm getting into, thank you very much.

**Comment by hazelnot** [comment]
AUR _packages_ aren't, the AUR itself is

**Comment by grapeli** [comment]
Maybe this ticket needs to be reopened.
I checked under Fedora 35 and there is also buffer overflow detected there. Of course, this is the case with dosbox-x hardening (-D_FORTIFY_SOURCE = 2).
https://developers.redhat.com/blog/2018/03/21/compiler-and-linker-flags-gcc/
Dosbox-staging compiled with the same flags works fine with c&c `setup.exe`.
```
CFLAGS="-march=x86-64 -mtune=native -O2 -msse2 -mfpmath=sse -mstackrealign -pipe -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection -fasynchronous-unwind-tables"
CXXFLAGS="-march=x86-64 -mtune=native -O2 -msse2 -mfpmath=sse -mstackrealign -pipe -fno-strict-aliasing -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection -fasynchronous-unwind-tables -Wp,-D_GLIBCXX_ASSERTIONS"
LDFLAGS="-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now"
```

I attach the full gdb log.
[gdb.backtrace.log](https://github.com/joncampbell123/dosbox-x/files/8132434/gdb.backtrace.log)

**Comment by maron2000** [comment]
There was a invalid value in strncpy.
`ctail.count` is too large in some case.
The following is a proposal of fix.
We have to check if `ctail.count` is used in other places of the code, and there may be other root causes of its invalid value.
But I confirmed C&C setup can start without `loadfix` or `minimum mcb segment` workarounds.
```cpp
// line 2147 of dos.cpp
memset(appname, 0, sizeof(appname));  // <-- add
memset(appargs, 0, sizeof(appargs));  // <-- add
strcpy(appname, name1);
//strncpy(appargs, ctail.buffer, ctail.count); <-- comment out
strncpy(appargs, ctail.buffer, strlen(ctail.buffer)); // <-- add
//*(appargs+ctail.count)=0; <-- comment out
```

**Comment by grapeli** [comment]
@maron2000 
I tested this patch and dosbox-x (security hardening build) works fine with the default settings after running `setup.exe` or `install.exe` from command & conquer.

Just to add that buffer overflow is detected when the core C library (glibc) is also built with `-D_FORTIFY_SOURCE=2`.
I tested under Debian 11.2 and its toolchain (glibc-2.31, gcc-10.2.1) does not detect a buffer overflow. It is not hardened enough.
